src/call_reply.c

```diff
@@ -0,0 +1,516 @@
+            freeCallReplyInternal(rep->val.array + i * 2 + 1);
+        }
+        zfree(rep->val.array);
+    }
+
+    if (rep->attribute) {
+        freeCallReplyInternal(rep->attribute);
+        zfree(rep->attribute);
+    }
+}
+
+/* Free the given call reply and its children (in case of nested reply) recursively.
+ * If private data was set when the CallReply was created it will not be freed, as it's
+ * the caller's responsibility to free it before calling freeCallReply(). */
+void freeCallReply(CallReply *rep) {
+    if (!(rep->flags & REPLY_FLAG_ROOT)) {
+        return;
+    }
+    if (rep->flags & REPLY_FLAG_PARSED) {
+        freeCallReplyInternal(rep);
+    }
+    sdsfree(rep->original_proto);
+    zfree(rep);
+}
+
+static const ReplyParserCallbacks DefaultParserCallbacks = {
+    .null_callback = callReplyNull,
+    .bulk_string_callback = callReplyBulkString,
+    .null_bulk_string_callback = callReplyNullBulkString,
+    .null_array_callback = callReplyNullArray,
+    .error_callback = callReplyError,
+    .simple_str_callback = callReplySimpleStr,
+    .long_callback = callReplyLong,
+    .array_callback = callReplyArray,
+    .set_callback = callReplySet,
+    .map_callback = callReplyMap,
+    .double_callback = callReplyDouble,
+    .bool_callback = callReplyBool,
+    .big_number_callback = callReplyBigNumber,
+    .verbatim_string_callback = callReplyVerbatimString,
+    .attribute_callback = callReplyAttribute,
+    .error = callReplyParseError,
+};
+
+/* Parse the buffer located in rep->original_proto and update the CallReply
+ * structure to represent its contents. */
+static void callReplyParse(CallReply *rep) {
+    if (rep->flags & REPLY_FLAG_PARSED) {
+        return;
+    }
+
+    ReplyParser parser = {.curr_location = rep->proto, .callbacks = DefaultParserCallbacks};
+
+    parseReply(&parser, rep);
+    rep->flags |= REPLY_FLAG_PARSED;
+}
+
+/* Return the call reply type (REDISMODULE_REPLY_...). */
+int callReplyType(CallReply *rep) {
+    if (!rep) return REDISMODULE_REPLY_UNKNOWN;
+    callReplyParse(rep);
+    return rep->type;
+}
+
+/* Return reply string as buffer and len. Applicable to:
+ * - REDISMODULE_REPLY_STRING
+ * - REDISMODULE_REPLY_ERROR
+ *
+ * The return value is borrowed from CallReply, so it must not be freed
+ * explicitly or used after CallReply itself is freed.
+ *
+ * The returned value is not NULL terminated and its length is returned by
+ * reference through len, which must not be NULL.
+ */
+const char *callReplyGetStr(CallReply *rep, size_t *len) {
+    callReplyParse(rep);
+    if (rep->type != REDISMODULE_REPLY_STRING &&
+        rep->type != REDISMODULE_REPLY_ERROR) return NULL;
+    if (len) *len = rep->len;
+    return rep->val.str;
+}
+
+/* Return a long long reply value. Applicable to:
+ * - REDISMODULE_REPLY_INTEGER
+ */
+long long callReplyGetLongLong(CallReply *rep) {
+    callReplyParse(rep);
+    if (rep->type != REDISMODULE_REPLY_INTEGER) return LLONG_MIN;
```

Comment:
This is confusing, because `rep->val.ll` is allowed to be `LLONG_MIN`, or whether it should be `assert`.

Reply:
@sundb this is how it is today, I did not change that : https://github.com/redis/redis/blob/1895e134a77efd789b1a6daee76a6ba5ec90e516/src/module.c#L3950

Reply:
if `rep->type` is `REDISMODULE_REPLY_STRING`, `callReplyGetLongLong(rep)` will return `LLONG_MIN`.
so how does the caller determine that it returned a error? `if(callReplyGetLongLong(rep) == LLONG_MIN)`?

Reply:
I understand what you are saying, I just mentioned that it is already like this today (and it's not like this PR breaks it). Also, the user can check the reply type to understand if it's long and not count on the result of `callReplyGetLongLong`.

Reply:
Summarizing:
1. On Lua this is a dead code because Lua will never call this function when the reply is not an integer
2. Modules can do whatever they want but if the module reaches this code then it must be a bug in the module, the module has a way to check the reply type before calling this function.

I believe the correct way was to assert that the type is REDISMODULE_REPLY_INTEGER, it will hold for Lua but it might crash modules that misuse the API it might be considered a breaking change. There is no good way to return an error to the module (with the current API),  so probably the best thing to do is to keep the behavior as it is today.